File: c/telegram_api_test/ssl_test.c
#include <stdio.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <netdb.h>
#include <unistd.h>
#include <string.h>


#define TOKEN "7822096798:AAEBztEDKoWcYQnn-UgEBu8Kn7Uiywr3dmk"
#define CHAT_ID "719439900"

#define REQUEST "GET /bot%s/%s HTTP/1.1\r\nContent-Type: application/json\r\n"\
    "Host: api.telegram.org\r\n"\
    "{\"chat_id\": \"%s\", \"text\": \"%s\"}"

#define TEST_REQUEST "POST /bot7822096798:AAEBztEDKoWcYQnn-UgEBu8Kn7Uiywr3dmk/sendMessage HTTP/1.1\r\n"\
    "Host: api.telegram.org\r\nContent-Type: application/json\r\nContent-Length: %d\r\n\r\n%s"

#define CONTENT  "{\"chat_id\": \"719439900\", \"text\": \"Greatings!!!\"}"

int main() {

    const char *message = "Hello my dear friends!!!";
    char request_buf[1024] = {0};

    OPENSSL_init_ssl(0, NULL);

    SSL_CTX *ctx = SSL_CTX_new(TLS_client_method());

    if (ctx == NULL) {
        fprintf(stderr, "SSL_CTX_new() error\n");
        return 1;
    }
    struct addrinfo hints = {0}, *res;
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;

    if (getaddrinfo("api.telegram.org", "443", &hints, &res) != 0) {
        fprintf(stderr, "getaddrinfo() error\n");
    }

    int sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
    if (sock < 1) {
        fprintf(stderr, "socket() error\n");
        return 1;
    }

    if (connect(sock, res->ai_addr, res->ai_addrlen) < 0) {
        fprintf(stderr, "connect() error\n");
        close(sock);
        return 1;
    }

    freeaddrinfo(res);

    SSL *ssl = SSL_new(ctx);
    if (ssl == NULL) {
        fprintf(stderr, "SSL_new() error\n");
        close(sock);
        return 1;
    }

    SSL_set_fd(ssl, sock);
    
    if (SSL_connect(ssl) <= 0) {
        ERR_print_errors_fp(stderr);
        close(sock);
        return 1;
    }

    //int message_len = sprintf(request_buf, REQUEST, TOKEN, "sendMessage", CHAT_ID, message);
    //int write_bytes = SSL_write(ssl, request_buf, strlen(request_buf));
    int content_len = strlen(CONTENT);
    int m_len = sprintf(request_buf, TEST_REQUEST, content_len, CONTENT);
    printf("%s\n", request_buf);
    int write_bytes = SSL_write(ssl, request_buf, strlen(request_buf));
    printf("Write bytes: %d\n", write_bytes);

    char response_buf[4096] = {0};
    int read_bytes = SSL_read(ssl, response_buf, 4096);
    printf("read bytes: %d\n", read_bytes);

    printf("%s\n", response_buf);

    SSL_shutdown(ssl);
    SSL_free(ssl);
    close(sock);
    SSL_CTX_free(ctx);

    return 0 ;
}

File: zig/test/token_test/ssl_test.zig
const std = @import("std");
const ssl = @cImport({
    @cInclude("openssl/ssl.h");
    @cInclude("openssl/err.h");
});
const net = @cImport({
    @cInclude("sys/socket.h");
    @cInclude("arpa/inet.h");
    @cInclude("netdb.h");
    @cInclude("unistd.h");
});

const HOST = "api.binance.com";
const PORT = "443";

const REQUEST = "GET /api/v3/ticker?symbol=BTCUSDT HTTP/1.1\r\nHost: api.binance.com\r\nConnection: keep-alive\r\n\r\n";

pub fn main() !void {
    // _ = ssl.SSL_library_init();
    _ = ssl.OPENSSL_init_ssl(0, null);
    // _ = ssl.SSL_load_error_strings();
    _ = ssl.OPENSSL_init_ssl(ssl.OPENSSL_INIT_LOAD_SSL_STRINGS | ssl.OPENSSL_INIT_LOAD_CRYPTO_STRINGS, null);
    // _ = ssl.OpenSSL_add_all_algorithms();
    _ = ssl.OPENSSL_init_crypto(ssl.OPENSSL_INIT_ADD_ALL_CIPHERS | ssl.OPENSSL_INIT_ADD_ALL_DIGESTS, null);
    const ctx = ssl.SSL_CTX_new(ssl.TLS_client_method().?) orelse {
        std.debug.print("SSL_CTX_new error\n", .{});
        return;
    };
    var hints = net.addrinfo{.ai_family = net.AF_INET, .ai_socktype = net.SOCK_STREAM};
    var res: *net.addrinfo = undefined;

    if (net.getaddrinfo(HOST, PORT, &hints, @ptrCast(&res)) != 0) {
        std.debug.print("getaddrinfo error\n", .{});
        return;
    }
    const sock = net.socket(res.*.ai_family, res.*.ai_socktype, res.*.ai_protocol);
    if (sock <= 0) {
        std.debug.print("socket error\n", .{});
        return;
    }

    if (net.connect(sock, res.*.ai_addr, res.*.ai_addrlen) != 0) {
        std.debug.print("connet error\n", .{});
        return;
    }
    net.freeaddrinfo(@ptrCast(res));

    const ssl_v = ssl.SSL_new(ctx).?;

    _ = ssl.SSL_set_fd(ssl_v, sock);

    if (ssl.SSL_connect(ssl_v) <= 0) {
        std.debug.print("SSL_connect error\n", .{});
        return;
    }

    const write_bytes = ssl.SSL_write(ssl_v, REQUEST, REQUEST.len);
    if (write_bytes <= 0) {
        std.debug.print("SSL_write error\n", .{});
        return;
    }

    var buf: [4096]u8 = .{0} ** 4096;

    const read_bytes = ssl.SSL_read(ssl_v, buf[0..].ptr, 4096);
    std.debug.print("Read {d} bytes\n", .{read_bytes});
    std.debug.print("{s}\n", .{buf});
    

    _ = ssl.SSL_shutdown(ssl_v);
    ssl.SSL_free(ssl_v);
    ssl.SSL_CTX_free(ctx);
    _ = net.close(sock);
}

